src/Data/Text/Internal/Search.hs

```diff
@@ -54,14 +54,19 @@ data T = {-# UNPACK #-} !Word64 :* {-# UNPACK #-} !Int
 indices :: Text                -- ^ Substring to search for (@needle@)
         -> Text                -- ^ Text to search in (@haystack@)
         -> [Int]
-indices (Text narr noff nlen)
+indices needle@(Text narr noff nlen)
   | nlen == 1 = scanOne (A.unsafeIndex narr noff)
   | nlen <= 0 = const []
-  | otherwise = scan
+  | otherwise = indices' needle
+{-# INLINE indices #-}
+
+-- | nlen must be >= 2, otherwise nindex causes access violation
+indices' :: Text -> Text -> [Int]
+indices' (Text narr noff nlen) (Text harr@(A.ByteArray harr#) hoff hlen) = loop (hoff + nlen)
   where
     nlast    = nlen - 1
     !z       = nindex nlast
-    nindex k = A.unsafeIndex narr (noff+k)
+    nindex k = if k < 0 then 0 else A.unsafeIndex narr (noff+k)
```

Comment:
Where's this check needed?

Reply:
Ah, it became redundant after I factored `indices'` out. Removed.